lib/rucio/daemons/reaper/reaper.py

```diff
@@ -217,6 +222,26 @@ def delete_from_storage(heartbeat_handler, hb_payload, replicas, prot, rse_info,
                 deletion_dict['duration'] = duration
                 add_message('deletion-failed', deletion_dict)
 
+            finally:
+                # this control loop will run indefinetely until all deletions have gone through (or failed).
+                # it assumes that for each deletion a timeout will occur (which will fail the deletion).
+                # if that assumption is not true, we need to introduce a maximum retry counter to avoid the worker hanging
+                # on individual deletions.
+                num_replicas_processed += 1
+
+                # after each replica is deleted we evaluate how much time we have left to delete.
+                # if we are not able to delete all the replicas we got in, other workers will take the replicas because their update time will
+                # be more than 10 minutes (refer to function list_and_mark_unlocked_replicas).
+                # After 8 minutes, and if we still have some replicas to process, we bump the replicas updated_at field to current time (with will delay the selectability
+                # by other workers by 8+ minutes).
+                elapsed_time = time.time() - refresh_start_time
+                if elapsed_time > trigger_time:  # 8 minutes have passed
+                    to_refresh = replicas[num_replicas_processed + 1]  # only refresh the replicas we have not yet deleted
+                    ok = refresh_replicas(rse_id=rse_id, replicas=to_refresh)
+                    if not ok:
+                        logger(logging.WARNING, "Failed to bump updated_at for replicas BEING_DELETED")
+                    refresh_start_time = time.time()  # reset it so we can trigger new refresh cycles.
```

Comment:
This check could also be done at the beginning of the loop, so that there's no need to even check if there still are replicas to process - as it's done at the beginning of the loop, we know that there's at least one more replica to process.